src/authorization/WebAclReader.ts

```diff
@@ -57,79 +62,56 @@ export class WebAclReader extends PermissionReader {
-      this.logger.debug(`Determining ${identifier.path} permissions requires verifying parent container permissions`);
-      const parentPermissions = acl.targetAcl === acl.parentAcl ?
-        permissions :
-        await this.findPermissions(acl.parentAcl!, credentials, false);
-
-      // https://solidproject.org/TR/2021/wac-20210711:
-      // When an operation requests to create a resource as a member of a container resource,
-      // the server MUST match an Authorization allowing the acl:Append or acl:Write access privilege
-      // on the container for new members.
-      permissions[CredentialGroup.agent]!.create = parentPermissions[CredentialGroup.agent]!.append;
-      permissions[CredentialGroup.public]!.create = parentPermissions[CredentialGroup.public]!.append;
-
-      // https://solidproject.org/TR/2021/wac-20210711:
-      // When an operation requests to delete a resource,
-      // the server MUST match Authorizations allowing the acl:Write access privilege
-      // on the resource and the containing container.
-      permissions[CredentialGroup.agent]!.delete =
-        permissions[CredentialGroup.agent]!.write && parentPermissions[CredentialGroup.agent]!.write;
-      permissions[CredentialGroup.public]!.delete =
-        permissions[CredentialGroup.public]!.write && parentPermissions[CredentialGroup.public]!.write;
-    }
-    return permissions;
+    this.logger.debug(`Retrieving permissions of ${credentials.agent?.webId ?? 'an unknown agent'}`);
+    const aclMap = await this.getAclMatches(new Set(accessMap.keys()));
+    const storeMap = await this.filterAclMap(aclMap);
+    return await this.findPermissions(storeMap, credentials);
   }
 
   /**
    * Finds the permissions in the provided WebACL quads.
-   * @param acl - Store containing all relevant authorization triples.
+   *
+   * Rather than restricting the search to only the required modes,
+   * we collect all modes in order to have complete metadata (for instance, for the WAC-Allow header).
+   *
+   * @param aclMap - A map containing stores of ACL data linked to their relevant identifiers.
    * @param credentials - Credentials to check permissions for.
-   * @param isAcl - If the target resource is an acl document.
    */
-  private async findPermissions(acl: Store, credentials: CredentialSet, isAcl: boolean): Promise<PermissionSet> {
-    const publicPermissions = await this.determinePermissions(acl, credentials.public);
-    const agentPermissions = await this.determinePermissions(acl, credentials.agent);
+  private async findPermissions(aclMap: Map<Store, ResourceIdentifier[]>, credentials: CredentialSet):
+  Promise<PermissionMap> {
+    const result: PermissionMap = new IdentifierMap();
+    for (const [ store, aclIdentifiers ] of aclMap) {
+      const publicPermissions = await this.determinePermissions(store, credentials.public);
+      const agentPermissions = await this.determinePermissions(store, credentials.agent);
```

Comment:
I'm confused by the question since the credentials are passed as a parameter. Unless the question is why this is not a single call of `const permissions = await this.determinePermissions(store, credentials);`? To which the main answer is that WebACL only supports those two specific credential groups so this makes it so only those two are used.